fetcher.go

```diff
@@ -20,11 +23,21 @@ type FetcherConfig struct {
 	blockService blockservice.BlockService
 }
 
-type Fetcher struct {
-	// TODO: for now, passing this to instantiation of block session is enough to
-	// cancel on session context cancel, but we may want to use this direct reference
-	// more tightly in this code
-	ctx         context.Context
+type Fetcher interface {
+	// NodeMatching traverses a node graph starting with the provided node using the given selector and possibly crossing
+	// block boundaries. Each matched node is sent to the FetchResult channel.
+	NodeMatching(ctx context.Context, node ipld.Node, match selector.Selector) (<-chan FetchResult, <-chan error)
+
+	// BlockOfType fetches a node graph of the provided type corresponding to single block by link.
+	BlockOfType(ctx context.Context, link ipld.Link, ptype ipld.NodePrototype) (ipld.Node, error)
+
+	// BlockMatchingOfType traverses a node graph starting with the given link using the given selector and possibly
+	// crossing block boundaries. The nodes will be typed using the provided prototype. Each matched node is sent to
+	// the FetchResult channel.
+	BlockMatchingOfType(ctx context.Context, root ipld.Link, match selector.Selector, ptype ipld.NodePrototype) (<-chan FetchResult, <-chan error)
```

Comment:
Should (both of) the result channels be expected to be closed when this request is done?

Will the error channel be able to yield more than one error?  Can FetchResult and errors be yielded intermingled?

Would this perhaps be clearer if this was unified into one channel, and FetchResult just contained either a Result or an error?

Reply:
You're last comment is correct. I gave to 2 channels a go, but I think it's just too complicated to explain the behavior and properly process the results. I've changed it to a single channel with an error in the result.

Reply:
Strike that. @hannahhoward pointed out https://github.com/ipfs/interface-go-ipfs-core/pull/62. The reasoning here makes sense, so I'm keeping the error channel, odifying it so that it's always buffered and closed on completion and then documenting the behavior in the interface.